src/lua/datetime.lua

```diff
@@ -420,6 +625,221 @@ local function datetime_tostring(self)
+        s = qtail(s) .. ('%+d hours, %d minutes, %s'):
+                        format(o.hour, math_fmod(o.min, 60),
+                               seconds(false, sec, nsec))
+    else
+        s = qtail(s) .. ('%+d days, %d hours, %d minutes, %s'):
+                        format(o.day, math_fmod(o.hour, 24),
+                               math_fmod(o.min, 60), seconds(false, sec, nsec))
+    end
+    return s
+end
+
+local function datetime_increment_by(self, direction, years, months,
+                                     seconds, nanoseconds, adjust)
+    -- operations with intervals should be done using local human dates
+    -- not UTC dates, thus normalize to local timezone before operations.
+    local dt = local_dt(self)
+    local secs, nsec = local_secs(self), self.nsec
+    local offset = self.tzoffset
+
+    local is_ym_updated = false
+    adjust = adjust or DEF_DT_ADJUST
+
+    if years ~= 0 then
+        check_range(years, MIN_DATE_YEAR, MAX_DATE_YEAR, 'years')
+        local approx_dt = dt + direction * years * AVERAGE_DAYS_YEAR
+        check_dt(approx_dt, direction, years, 'years')
+        -- tnt_dt_add_years() not handle properly DT_SNAP or DT_LIMIT mode
+        -- so use tnt_dt_add_months() as a work-around
+        dt = builtin.tnt_dt_add_months(dt, direction * years * 12, adjust)
+        is_ym_updated = true
+    end
+    if months ~= 0 then
+        local new_dt = dt + direction * months * AVERAGE_DAYS_MONTH
+        check_dt(new_dt, direction, months, 'months')
+        dt = builtin.tnt_dt_add_months(dt, direction * months, adjust)
+        is_ym_updated = true
+    end
+    if is_ym_updated then
+        secs = (dt - DAYS_EPOCH_OFFSET) * SECS_PER_DAY + secs % SECS_PER_DAY
+    end
+
+    if seconds ~= 0 then
+        secs = secs + direction * seconds
+    end
+
+    if nanoseconds ~= 0 then
+        nsec = nsec + direction * nanoseconds
+    end
+
+    secs, self.nsec = normalize_nsec(secs, nsec)
+    check_dt((secs + SECS_EPOCH_OFFSET) / SECS_PER_DAY, direction)
+    self.epoch = utc_secs(secs, offset)
+    return self
+end
+
+local function date_first(lhs, rhs)
+    if is_datetime(lhs) then
+        return lhs, rhs
+    else
+        return rhs, lhs
+    end
+end
+
+local function error_incompatible(name)
+    error(("datetime:%s() - incompatible type of arguments"):
+          format(name), 3)
+end
+
+--[[
+Matrix of subtraction operands eligibility and their result type
+
+|                 | datetime | interval | table    |
++-----------------+----------+----------+----------+
+| datetime        | interval | datetime | datetime |
+| interval        |          | interval | interval |
+| table           |          |          |          |
+]]
+local function datetime_interval_sub(lhs, rhs)
+    check_date_interval(lhs, "operator -")
+    check_date_interval_table(rhs, "operator -")
+    local left_interval = is_table(lhs) or is_interval(lhs)
+    local right_interval = is_table(rhs) or is_interval(rhs)
```

Comment:
Please, name flags using special prefix such as `is_`, `has_`, `does_`, etc. These variables are not intervals, they are flags.

Reply:
(it would be not prefix in this case)

Reply:
Like that...
```diff
--- a/src/lua/datetime.lua
+++ b/src/lua/datetime.lua
@@ -750,23 +750,23 @@ Matrix of subtraction operands eligibility and their result type
 local function datetime_interval_sub(lhs, rhs)
     check_date_interval(lhs, "operator -")
     check_date_interval_table(rhs, "operator -")
-    local left_interval = is_table(lhs) or is_interval(lhs)
-    local right_interval = is_table(rhs) or is_interval(rhs)
+    local left_is_interval = is_table(lhs) or is_interval(lhs)
+    local right_is_interval = is_table(rhs) or is_interval(rhs)
     local year, month, secs, nsec, adjust
 
     -- left is date, right is interval
-    if not left_interval and right_interval then
+    if not left_is_interval and right_is_interval then
         year, month, secs, nsec, adjust = interval_decouple_args(rhs)
         return datetime_increment_by(datetime_new_copy(lhs), -1, year, month,
                                      secs, nsec, adjust)
     -- left is date, right is date
-    elseif not left_interval and not right_interval then
+    elseif not left_is_interval and not right_is_interval then
         local obj = interval_new()
         obj.sec, obj.nsec = normalize_nsec(lhs.epoch - rhs.epoch,
                                            lhs.nsec - rhs.nsec)
         return obj
     -- both left and right are intervals
-    elseif left_interval and right_interval then
+    elseif left_is_interval and right_is_interval then
         year, month, secs, nsec, adjust = interval_decouple_args(lhs)
         local obj = interval_init(year, month, secs, nsec, adjust)
         year, month, secs, nsec = interval_decouple_args(rhs)
@@ -793,17 +793,17 @@ local function datetime_interval_add(lhs, rhs)
 
     check_date_interval(lhs, "operator +")
     check_interval_table(rhs, "operator +")
-    local left_interval = is_table(lhs) or is_interval(lhs)
-    local right_interval = is_table(rhs) or is_interval(rhs)
+    local left_is_interval = is_table(lhs) or is_interval(lhs)
+    local right_is_interval = is_table(rhs) or is_interval(rhs)
     local year, month, secs, nsec, adjust
 
     -- left is date, right is interval
-    if not left_interval and right_interval then
+    if not left_is_interval and right_is_interval then
         local obj = datetime_new_copy(lhs)
         year, month, secs, nsec, adjust = interval_decouple_args(rhs)
         return datetime_increment_by(obj, 1, year, month, secs, nsec, adjust)
     -- both left and right are intervals
-    elseif left_interval and right_interval then
+    elseif left_is_interval and right_is_interval then
         year, month, secs, nsec, adjust = interval_decouple_args(lhs)
         local obj = interval_init(year, month, secs, nsec, adjust)
         year, month, secs, nsec = interval_decouple_args(rhs)
```

Because making `is_` prefix here will change make it having incorrect or meaningless semantics `is_left_interval` or `is_right_interval`.